scripts/stage-docs.sh

```diff
@@ -0,0 +1,162 @@
+ -d dry-run mode
+ -h this help message
+EOF
+      exit 1
+    ;;
+    \?) echo "Invalid option -$opt ${OPTARG}" >&2
+    exit 1
+    ;;
+  esac
+
+done
+
+if [[ -z ${RELEASE_VERSION} ]]; then
+  echo "No version specified, aborting"
+  exit 1
+fi
+
+if [[ -z ${WEBSITE_REPO_URL} ]]; then
+  echo "No website repository URL specified, aborting"
+  exit 1
+fi
+
+BLUE='\033[0;34m'
+NC='\033[0m' # No Color
+
+ORIGINAL_WORKING_BRANCH=$(git branch --show-current)
+
+cleanup() {
+    if [[ -n ${ORIGINAL_WEBSITE_WORKING_BRANCH} ]]; then
+        git checkout "${ORIGINAL_WEBSITE_WORKING_BRANCH}" || true
+    fi
+
+    if [[ ${ORIGINAL_GH_DEFAULT_REPO} ]]; then
+      gh repo set-default "${ORIGINAL_GH_DEFAULT_REPO}"
+    fi
+
+    # Note that git branch -D echos the sha of the deleted branch to
+    # stdout.  This is great for debugging the release process as it
+    # lets the developer restore to the state of the tree.
+    if [[ "${DRY_RUN:-false}" == true && ${RELEASE_DOCS_BRANCH} ]]; then
+        git branch -D "${RELEASE_DOCS_BRANCH}" || true
+    fi
+
+    cd ../kroxylicious/
+
+    if [[ -n ${ORIGINAL_WORKING_BRANCH} ]]; then
+        git checkout "${ORIGINAL_WORKING_BRANCH}" || true
+    fi
+}
+
+trap cleanup EXIT
+
+git stash --all
+
+RELEASE_DATE=$(date -u '+%Y-%m-%d')
+RELEASE_TAG="v${RELEASE_VERSION}"
+RELEASE_DOCS_BRANCH="prepare-${RELEASE_TAG}-release-docs-${RELEASE_DATE}"
+
+# Use a `/.` at the end of the source path to avoid the source path being appended to the destination path if the `.../_files/` folder already exists
+KROXYLICIOUS_DOCS_LOCATION="../kroxylicious/docs/."
+WEBSITE_DOCS_LOCATION="./docs/${RELEASE_TAG}"
+
+if [[ "${DRY_RUN:-false}" == true ]]; then
+    #Disable the shell check as the colour codes only work with interpolation.
+    # shellcheck disable=SC2059
+    printf "${BLUE}Dry-run mode:${NC} no remote tags or PRs will be created, artefacts will be deployed to: ${DEPLOY_DRY_RUN_DIR}\n"
+    GIT_DRYRUN="--dry-run"
+fi
+
+echo "Checking out tags/${RELEASE_TAG} in  in $(git remote get-url "${REPOSITORY}")"
+git checkout "tags/${RELEASE_TAG}"
+
+# Move up a directory so we don't end up with website files in the main repository
+cd ../
+echo "In '$(pwd)', cloning website repository at ${WEBSITE_REPO_URL}"
+git clone "${WEBSITE_REPO_URL}"
+cd ./kroxylicious.github.io/
+
+ORIGINAL_WEBSITE_WORKING_BRANCH=$(git branch --show-current)
+
+echo "Creating branch ${RELEASE_DOCS_BRANCH} from ${BRANCH_FROM} in $(git remote get-url "${REPOSITORY}")"
+git checkout -b "${RELEASE_DOCS_BRANCH}"
+
+echo "Copying release docs from ${KROXYLICIOUS_DOCS_LOCATION} to ${WEBSITE_DOCS_LOCATION}/_files"
+mkdir -p "${WEBSITE_DOCS_LOCATION}/"
+cp -R "${KROXYLICIOUS_DOCS_LOCATION}" "${WEBSITE_DOCS_LOCATION}/_files"
+# Remove README.md from copied files
+rm "${WEBSITE_DOCS_LOCATION}/_files/README.md"
+
+echo "Creating AsciiDoc entrypoint file at ${WEBSITE_DOCS_LOCATION}/index.adoc"
+RELEASE_DOCS_INDEX_TEMPLATE="
```

Comment:
It would, and I did consider that option, but it complicates things in a few ways:
1. We'd have to keep that template somewhere that it wouldn't get seen and rendered by the AsciiDoc plugin, while also preventing Jekyll from trying to serve it as a raw asset, because when Jekyll encounters any file that isn't markdown it will attempt to publish it raw unless it's specifically excluded in config.
   - This is doable but a bit of a pain to set up.
2. We then have to parse and modify that file using a tool like `sed` which is a lot harder to read than the simple variable substitution in the string that I went with in the end. It also (marginally) impacts the speed and complexity of the script as it means reading in one file, modifying the contents in memory, and then outputting that data into another file.
   - This is also doable, but I'm not sure it's worth the effort of saving a few lines (or possibly not saving any lines at all, if I'm really bad at regex 😆 ).
3. There's also the possibility that having the template in another repository could make this pipeline a bit fragile, because if that template file doesn't exist in the website repository (i.e. it gets deleted or moved for whatever reason) then the script could (and probably should) fail, which would mean that the later steps in the pipeline would (at present, with how `stage_release.yaml` is set up) treat it as the _release_ failing, when in actuality the issue is a change to the website.
   - There's probably a way to minimise how fragile this makes the pipeline (i.e. we might be able to change how the `Drop stage repository, if failed` step checks for a failure so it's more specific), but I think the easiest way to circumvent this problem is to just leave it as it is, keeping this script together with any files/contents it needs to be able to run. I'm not sure about making the release pipeline for this repository dependent on files in other repositories, when those repositories aren't static and those files aren't actually critical to the release.

I'm happy to go with a template file in the website repo if we feel that's the best option, but I'm just not sure it's worth the time to do so.

Reply:
Happy with your approach.
btw.  on point 2), I would have probably used `envsubst`.